pydantic_ai_graph/pydantic_ai_graph/graph.py

```diff
@@ -2,169 +2,232 @@
 
 import base64
 import inspect
-import types
-from dataclasses import dataclass
-from datetime import datetime, timezone
+from dataclasses import dataclass, field
 from pathlib import Path
 from time import perf_counter
 from typing import TYPE_CHECKING, Any, Generic
 
 import logfire_api
-from typing_extensions import TypeVar, assert_never
+from typing_extensions import Never, ParamSpec, Protocol, TypeVar, assert_never
 
 from . import _utils
-from .nodes import BaseNode, End, GraphContext, GraphOutputT, NodeDef
-from .state import Snapshot, StateT
+from ._utils import get_parent_namespace
+from .nodes import BaseNode, End, GraphContext, NodeDef
+from .state import EndEvent, GraphHistoryItem, StateT, Step
 
-__all__ = ('Graph',)
+__all__ = ('Graph', 'GraphRun', 'GraphRunner')
 
 _logfire = logfire_api.Logfire(otel_scope='pydantic-ai-graph')
-GraphInputT = TypeVar('GraphInputT', default=Any)
+
+RunSignatureT = ParamSpec('RunSignatureT')
+RunEndT = TypeVar('RunEndT', default=None)
+NodeRunEndT = TypeVar('NodeRunEndT', covariant=True, default=Never)
+
+
+class StartNodeProtocol(Protocol[RunSignatureT, StateT, NodeRunEndT]):
+    def get_id(self) -> str: ...
+    def __call__(self, *args: RunSignatureT.args, **kwargs: RunSignatureT.kwargs) -> BaseNode[StateT, NodeRunEndT]: ...
 
 
-# noinspection PyTypeHints
 @dataclass(init=False)
-class Graph(Generic[StateT, GraphInputT, GraphOutputT]):
+class Graph(Generic[StateT, RunEndT]):
     """Definition of a graph."""
 
-    first_node: NodeDef[StateT, Any, Any]
-    nodes: dict[str, NodeDef[StateT, Any, Any]]
-    name: str | None
+    name: str
+    nodes: tuple[type[BaseNode[StateT, RunEndT]], ...]
+    node_defs: dict[str, NodeDef[StateT, RunEndT]]
 
     def __init__(
         self,
-        first_node: type[BaseNode[StateT, GraphInputT, GraphOutputT]],
-        *other_nodes: type[BaseNode[StateT, Any, GraphOutputT]],
-        name: str | None = None,
+        name: str = 'graph',
```

Comment:
we'll want the same logic as on agents where we infer the name from the frame, hence this should be be `None` by default.